### PR TITLE
Use on-exit-leak-free for transport on-exit

### PR DESCRIPTION
**Pull request by mcollina**
Uses WeakRef, WeakMap and FinalizationGroup to implement a leak-free exit listener.

### CODE REVIEW

File: lib/transport.js [hunk 1]
```diff
@@ -6,6 +6,27 @@ const { join, isAbsolute } = require('path')
 
 const ThreadStream = require('thread-stream')
 
+function setupOnExit (stream) {
+  /* istanbul ignore next */
+  if (global.WeakRef && global.WeakMap && global.FinalizationRegistry) {
```
Comment (jsumners) [inline lib/transport.js:11]: Looks like v14.17.3 is the first version with full support for all of these. Do we need to make this clear in the documentation?

Comment (davidmarkclements) [reply lib/transport.js:11]: can we just release 7 for min Node version 14.7.3?

Comment (davidmarkclements) [reply lib/transport.js:11]: I say remove the checks and do a version check in index.js, throw and say use Pino 6 for older node versions

**Review by davidmarkclements — CHANGES_REQUESTED** [review]
no need for the checks imo, just drop support for older node in v7

**Comment by mcollina** [comment]
We need to support Node.js v12, it's still LTS and we do not want to force one of our consumers (Fastify) to drop it.

**Review by davidmarkclements — APPROVED** [review]
my bad, I forgot 12 has a long life left

File: docs/api.md [hunk 2]
```diff
@@ -939,6 +939,10 @@ const transports = pino.transport({
 pino(transports)
 ```
 
+If `WeakRef`, `WeakMap` and `FinalizationRegistry` are available in the current runtime (v14.5.0+), then the thread
+will be automatically terminated in case the stream or logger goes out of scope.
+The `transport()` function adds a listener to `process.on('exit')` to ensure the worker is flushed and all data synced
+before the process exits.
```
Comment (mcollina) [inline docs/api.md:945]: @jsumners docs added

**Comment by jsumners** [comment]
> my bad, I forgot 12 has a long life left

"long" is a bit generous. EOL is April 2022.

**Review by jsumners — APPROVED** [review]
Looks good to me.

**Pull request merged**